Add state that apply DSA-4371-1 fix safely

marmarek opened the pull request:
Update libapt-pkg* with redirects disabled.

QubesOS/qubes-issues#4752

marmarek committed:
Add state that apply DSA-4371-1 fix safely

Update libapt-pkg* with redirects disabled.

QubesOS/qubes-issues#4752

marmarek committed:
Include update.vm state in the package.

This is the first state shipped by this package, so adjust FORMULA file
accordingly.

QubesOS/qubes-issues#4752

HW42 commented on update/vm.sls:
For me it's not clear in which order `refresh`, `dist_upgrade` and `require` are executed. I assume you have checked that it will be run in the correct order (refresh, require, upgrade).

marmarek replied:
`require` apply to the whole `pkg.uptodate` state, it's handled by salt logic, not actual `pkg.uptodate` function.

HW42 commented on update/vm.sls:
In my bash script I do a update run with an empty source.list first to ensure that all old list files get cleaned up. AFAIK, on network errors `apt-get update` keeps the old list files. OTOH probably nobody has `cdn-fastly.deb.debian.org` in their source list (the probably use `deb.debian.org` instead), so there should be no old lists available.

marmarek replied:
Does it matter? We check for explicit package version, not "the latest". So, on network error, the state will fail to find the right version (I've checked by putting 1.4.10 there).

HW42 replied:
Yes, AFAIU this bug can also be used on other downloaded files. So you can replace the `Packages` list with another version containing another hash for the "right" version.

marmarek replied:
`apt clean` call before?

HW42 replied:
`apt clean` does not remove the package lists (you normally need them). AFAIK you need to run apt-get update with empty sources.list or manually delete stuff in `/var/lib/apt/lists`.

HW42 commented:
What happens if the fixed package is already installed? AFAIU this will still change sources.list, update and then check the package version and then delete the sources.list.

marmarek commented:
Yes, it will do that...

HW42 commented:
Where is the normal update/upgrade done (after the buggy package has been updated, then again with the normal sources config).

marmarek commented:
The `pkg.installed` is about updating apt, the `pkg.uptodate` is about normal update.

marmarek commented:
Maybe better have uniform solution and call your script here? It isn't exactly declarative, but this one isn't either (because the need to create + remove sources list).

HW42 commented:
> The `pkg.installed` is about updating apt, the `pkg.uptodate` is about normal update.

So the order of the rm and pkg.uptodate is ensured?

> Maybe better have uniform solution and call your script here? It isn't exactly declarative, but this one isn't either (because the need to create + remove sources list).

Whatever you think works better. I think the problem with doing it every time needs to be fixes since refreshing the packages list can be rather slow. The list cleanup is, as noted above, rather theoretical since the user need to have a rather unusual source config to be affected (and then the attacker still needs to get the timing right).

marmarek commented:
> So the order of the rm and pkg.uptodate is ensured?

In theory no, (indeed `update` should require `rm`, not `{{pkg}}`), in practice yes as salt order states according to file order if not specified otherwise. But I don't think it's guaranteed this way.

marmarek commented:
I'm also slightly worried about name conflicts here - the state file is named `update/vm.sls`, which needs to be unique, regardless of its location. It isn't unthinkable that some users have already created such file on their system...

marmarek committed:
Use dsa-4371-update script instead of non-declarative states

The DSA-4371 update procedure, as written with plain states is not
really declarative. If for nothing else, apt configuration is created
and then removed, which also require strict state ordering to work.
Since @HW42 created a script for this anyway, lets use this script here
too.

QubesOS/qubes-issues#4752

marmarek committed:
Rename update/vm.sls to update/qubes-vm.sls to avoid conflicts

This name needs to be unique among all the states. update/vm.sls may be
too common (especially as it was recommended creating similar file on
mailing lists), rename it to update/qubes-vm.sls.

QubesOS/qubes-issues#4752
